File: runtime-py/src/bantamkit/workplan.py
"""Layer 1 — a deterministic work-plan planner: turn `depends_on` into batches.

`assets/schemas/shiftwork-checkpoint.json` has carried `plan.units[].depends_on` since
the contract was written and neither runtime has ever read it: cursor advance is
v1-linear in both, so every job this repo ran was executed as a straight line whatever
its dependency graph said. Measured 2026-09-18 over the 19 checkpoints in `.shiftwork/`,
215 units collapse to 119 batches — 96 serial steps (44.7 %) were false serialization.

This module is the mechanism underneath that. It is **Layer 1 (Core)** per
docs/architecture.md: a pure function, no filesystem, no clock, no `mcp` import, no
dependency outside the standard library. The shift-work adapter (Layer 5) and the tool
assets (Layer 2) are separate units in separate files, because a change lives in exactly
one layer.

The semantics are the documented defaults of `ex-flow@1.1.0` — the user's own MIT
package, which is what made the measurement above possible. It is the SPECIFICATION and
deliberately NOT a dependency: `runtime-ts/package.json` declares exactly one runtime
dependency, and `docs/porting.md` uses "no runtime dependency is allowed into
`runtime-ts`" as the live justification for three divergence rows. Taking `ex-flow`
would collapse that argument to save a hundred lines of graph code, and `runtime_py`
could not import it in any case. So both runtimes hand-write it and the conformance
suite proves they agreed. Design:
`docs/superpowers/specs/2026-09-18-workplan-dag-design.md`.

Determinism is not a nicety here — it is the whole gate. Two implementations that both
"topologically sort correctly" can still disagree about the order inside a batch and
about WHICH cycle they name in a graph that has two, and either disagreement turns the
conformance suite red for a reason that is not a defect. Hence the two ordering rules
below are stated as contract, not as an artifact of whatever the dict happened to do.
"""

from typing import Any

__all__ = ["plan"]


def _cycle_path(order: list[str], deps: dict[str, list[str]], pending: set[str]) -> list[str]:
    """Name one cycle among the nodes Kahn could not emit, reproducibly in any language.

    A graph can hold several cycles and every one of them is an equally correct answer,
    so "find a cycle" is not a specification — two runtimes would name two different
    ones and the differential would go red over nothing. The rule is therefore fixed:
    scan the un-emitted nodes in INPUT order, walk `depends_on` in DECLARED order, and
    the first repeated id closes the cycle. First node scanned, first edge walked, first
    repeat wins.

    Only un-emitted dependencies are walked. An id Kahn already emitted sits in a batch
    below and can never be part of a cycle, so following one would be a dead end rather
    than a different answer.

    The walk may begin at a node that is merely downstream of a cycle rather than in one
    (`a -> b -> c -> b`: `a` cannot be emitted, but `a` is not in the cycle). What is
    returned is the CYCLE, `b -> c -> b`, trimmed at the first occurrence of the repeated
    id — the tail that only led into it is not part of it.
    """
    start = next(node_id for node_id in order if node_id in pending)
    path = [start]
    seen = {start: 0}
    current = start
    while True:
        nxt = next((dep for dep in deps[current] if dep in pending), None)
        if nxt is None:
            # Unreachable for a graph Kahn stalled on: a node is only left pending
            # because at least one dependency of it is also still pending.
            return path + [path[0]]
        if nxt in seen:
            return path[seen[nxt]:] + [nxt]
        seen[nxt] = len(path)
        path.append(nxt)
        current = nxt


def plan(nodes: list[dict[str, Any]]) -> dict[str, Any]:
    """Batch a dependency graph into the levels that may run in parallel.

    nodes: [{"id": str, "depends_on": list[str], "priority": int}] — `depends_on`
    defaults to empty and `priority` defaults to 0, because the shift-work checkpoint
    schema has no priority field and this design does not add one.

    Returns either

        {"batches": [[id, ...], ...], "sequence": [id, ...], "width": int}

    where batch *k* holds every node whose dependencies all appear in batches < *k*,
    `sequence` is those batches flattened in order, and `width` is the largest batch
    length (the widest fan-out, which is the number an orchestrator needs to decide
    whether it can afford the batch) — or a structured refusal

        {"result": "error", "reason": str}

    which is the shape `shiftwork.py` already returns, so nothing new is invented for it.

    **Ordering inside a batch is priority descending, then insertion order** — the order
    the nodes were given, which for the shift-work adapter is `plan.units` order. Both
    halves are contract: the second is what makes two independent implementations agree.

    `plan([])` is `{"batches": [], "sequence": [], "width": 0}`. An empty plan is an
    ANSWER, not a refusal — the same reasoning `shiftwork.clock_in` already applies when
    its all-terminal test reports success instead of escalating on a dangling cursor.

    The three refusals are checked in a fixed order — duplicate id, then unknown
    dependency, then cycle — so a graph carrying two faults always names the same one.
    The later checks would be meaningless on the earlier faults' input anyway: a
    duplicate id makes "which node declares this" ambiguous, and an edge into a node
    that does not exist is not a cycle.
    """
    order: list[str] = []
    deps: dict[str, list[str]] = {}
    priority: dict[str, int] = {}

    for node in nodes:
        node_id = node["id"]
        if node_id in deps:
            return {"result": "error", "reason": f"duplicate node id {node_id}"}
        order.append(node_id)
        deps[node_id] = list(node.get("depends_on") or [])
        priority[node_id] = node.get("priority", 0)

    for node_id in order:
        for dep in deps[node_id]:
            if dep not in deps:
                return {
                    "result": "error",
                    "reason": f"node {node_id} depends on {dep}, which no node declares",
                }

    # Insertion index is the tie-break, so it is carried explicitly rather than being
    # left to the incidental order of a set or dict during the sweep below.
    index = {node_id: i for i, node_id in enumerate(order)}

    pending = set(order)
    emitted: set[str] = set()
    batches: list[list[str]] = []

    while pending:
        ready = [
            node_id
            for node_id in order
            if node_id in pending and all(dep in emitted for dep in deps[node_id])
        ]
        if not ready:
            path = _cycle_path(order, deps, pending)
            return {
                "result": "error",
                "reason": "the graph has a cycle: " + " -> ".join(path),
            }
        ready.sort(key=lambda node_id: (-priority[node_id], index[node_id]))
        batches.append(ready)
        pending.difference_update(ready)
        emitted.update(ready)

    sequence = [node_id for batch in batches for node_id in batch]
    return {
        "batches": batches,
        "sequence": sequence,
        "width": max((len(batch) for batch in batches), default=0),
    }

File: runtime-ts/src/workplan.ts
/**
 * Layer 1 — the work-plan planner: `depends_on` turned into batches that may run in parallel.
 *
 * A pure function. No filesystem, no clock, no `mcp` import, no dependency: `docs/architecture.md`
 * puts a deterministic graph planner in Core, and `shiftwork.ts` — which is Layer 5 — is the thing
 * that will call it, not the thing it lives in.
 *
 * WHERE THE SEMANTICS COME FROM. `docs/superpowers/specs/2026-09-18-workplan-dag-design.md`. The
 * user's own `ex-flow@1.1.0` measured the numbers in that spec, and it is deliberately NOT an
 * install: `runtime-ts/package.json` declares exactly one runtime dependency,
 * `@modelcontextprotocol/sdk`, and `docs/porting.md` uses "no runtime dependency is allowed into
 * `runtime-ts`" as the live justification for three divergence rows. So `ex-flow` is the
 * SPECIFICATION and this is one of two hand-written implementations held to it — the other being
 * `runtime-py/src/bantamkit/workplan.py`, written independently so that
 * `tools/conformance/suites/workplan.mjs` proves an agreement rather than a copy.
 *
 * WHAT IS DELIBERATELY LEFT OUT of `ex-flow`'s surface: `deadlineStrategy`, `weightStrategy`,
 * `fairnessPolicy`/`maxDeferralRounds` aging, `schedulerMode: "throughput"`, `concurrencyCap`,
 * `resourceCaps`, the clone modes and the config builder. None of the 19 measured checkpoints
 * supplies a deadline, a weight or a resource class, so every one of them would be untested
 * surface. `sequence` is `ex-flow`'s `fullSequence` under a shorter name.
 *
 * WHAT THIS DOES NOT CLAIM. `depends_on` encodes logical order, not file contention. Measured on
 * `.shiftwork/checkpoint-job44.json`, whose first batch is twelve units wide, seven of those
 * twelve name `runtime-ts/src/docread.ts`. The graph says they may run in parallel; the filesystem
 * does not. This function reports what the graph permits and the orchestrator stays responsible
 * for what it actually dispatches.
 */

/** One task. `priority` is optional and absent means `0` — the checkpoint schema has no such field. */
export interface PlanNode {
  id: string;
  depends_on: string[];
  priority?: number;
}

/** A plan: the batches, the same ids read end to end, and the widest fan-out in the plan. */
export interface PlanOk {
  batches: string[][];
  sequence: string[];
  width: number;
}

/**
 * A refusal, in the `{result, reason}` shape `shiftwork` already returns, so nothing new is
 * invented for it. The three sentences are pinned per-side in `test/workplan.test.mjs` because a
 * sentence both runtimes copied is invisible to the differential.
 */
export interface PlanError {
  result: 'error';
  reason: string;
}

export type PlanResult = PlanOk | PlanError;

const refuse = (reason: string): PlanError => ({ result: 'error', reason });

/**
 * Name the cycle that stopped Kahn, deterministically.
 *
 * Two runtimes left to their own devices will name two different — equally correct — cycles in the
 * same graph, and the conformance suite goes red for a reason that is not a defect. The rule, and
 * the three rulings that disambiguate it:
 *
 * - Scan the nodes Kahn could not emit **in input order** and start at the first.
 * - Walk `depends_on` **in declared order**, following only dependencies that are themselves
 *   UN-EMITTED. A dependency already emitted in an earlier batch can never close a cycle, so
 *   following it would dead-end on a graph that does have one. (Every un-emitted node has at least
 *   one un-emitted dependency — otherwise it was ready — so the walk can never run out of edges.)
 * - The first repeated id closes the cycle, and the reported path is the CYCLE, not the whole walk:
 *   the scan may have started merely downstream of the loop, and a sentence that named a node
 *   outside the loop would be describing something other than the cycle it claims to name.
 */
function cyclePath(nodes: readonly PlanNode[], emitted: ReadonlySet<string>): string[] {
  const byId = new Map(nodes.map((node) => [node.id, node] as const));
  const path: string[] = [];
  /** id -> where it sits in `path`, which is what turns "seen before" into "trim the tail here". */
  const at = new Map<string, number>();

  let current = nodes.find((node) => !emitted.has(node.id));
  while (current !== undefined) {
    const repeat = at.get(current.id);
    if (repeat !== undefined) return [...path.slice(repeat), current.id];
    at.set(current.id, path.length);
    path.push(current.id);
    const next = current.depends_on.find((dep) => !emitted.has(dep));
    current = next === undefined ? undefined : byId.get(next);
  }
  // Unreachable on a graph that reached here: the walk only ends by repeating an id.
  return path;
}

/**
 * Level-synchronous Kahn: batch *k* holds every node whose dependencies all appear in batches < *k*.
 *
 * **Empty input is not an error.** `plan([])` is `{batches: [], sequence: [], width: 0}` — a plan
 * with nothing left to do is the answer, the same reasoning `clock_in` applies when an all-terminal
 * checkpoint reports success rather than escalating on a dangling cursor.
 *
 * **The three refusals are checked in a fixed order** — duplicate id, then unknown dependency, then
 * cycle — because a graph can carry two faults at once and both runtimes have to name the same one.
 */
export function plan(nodes: readonly PlanNode[]): PlanResult {
  // 1 — duplicate id. First in input order wins.
  const declared = new Set<string>();
  for (const node of nodes) {
    if (declared.has(node.id)) return refuse(`duplicate node id ${node.id}`);
    declared.add(node.id);
  }

  // 2 — a dependency nothing declares. First node in input order, first dep in declared order.
  for (const node of nodes) {
    for (const dep of node.depends_on) {
      if (!declared.has(dep)) {
        return refuse(`node ${node.id} depends on ${dep}, which no node declares`);
      }
    }
  }

  // 3 — the batches.
  const emitted = new Set<string>();
  const batches: string[][] = [];
  const sequence: string[] = [];
  let width = 0;

  while (emitted.size < nodes.length) {
    /**
     * Carry the input index explicitly rather than leaning on `Array.prototype.sort` being stable.
     * It is, in every Node this package supports — but the tie-break is the thing the conformance
     * gate compares against Python, and a reader has to be able to SEE that it is insertion order
     * and not a property of the sort.
     */
    const ready: { id: string; priority: number; index: number }[] = [];
    for (const [index, node] of nodes.entries()) {
      if (emitted.has(node.id)) continue;
      if (node.depends_on.every((dep) => emitted.has(dep))) {
        ready.push({ id: node.id, priority: node.priority ?? 0, index });
      }
    }

    // Nothing is ready and something is left: the remainder contains a cycle.
    if (ready.length === 0) {
      return refuse(`the graph has a cycle: ${cyclePath(nodes, emitted).join(' -> ')}`);
    }

    // Priority descending, then insertion order.
    ready.sort((a, b) => b.priority - a.priority || a.index - b.index);

    const batch = ready.map((entry) => entry.id);
    batches.push(batch);
    for (const id of batch) {
      emitted.add(id);
      sequence.push(id);
    }
    if (batch.length > width) width = batch.length;
  }

  return { batches, sequence, width };
}
